File: QRCode/qr_code.cc
#include <opencv2/objdetect.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <iostream>

using namespace cv;
using namespace std;

void display(Mat &im, Mat &bbox)
{
  int n = bbox.rows;
  for (int i = 0; i < n; i++)
  {
    line(im,
         Point2i(bbox.at<float>(i, 0), bbox.at<float>(i, 1)),
         Point2i(bbox.at<float>((i + 1) % n, 0), bbox.at<float>((i + 1) % n, 1)),
         Scalar(255, 0, 0), 3);
  }
  imshow("Results", im);
}

int main(int argc, char *argv[])
{
  Mat inputImage;
  if (argc > 1)
    inputImage = imread(argv[1]);
  else
    inputImage = imread("../images/convoluted-webpage.png");

  QRCodeDetector qrDecoder;
  Mat bbox, rectifiedImage;

  std::string data = qrDecoder.detectAndDecode(inputImage, bbox, rectifiedImage);
  if (data.length() > 0)
  {
    cout << "Decoded Data : " << data << endl;
    display(inputImage, bbox);
    rectifiedImage.convertTo(rectifiedImage, CV_8UC3);
    imshow("Rectified QRCode", rectifiedImage);
  }
  else
  {
    cout << "QR Code not detected" << endl;
    imshow("Results", inputImage);
  }
  waitKey(0);
}

File: QRCode/qr_code.py
import cv2
import numpy as np
import sys
import time


def display(im, bbox):
    n = len(bbox)
    for j in range(n):
        cv2.line(im,
                 tuple(bbox[j][0]),
                 tuple(bbox[(j+1) % n][0]),
                 (255, 0, 0), 3)
        cv2.imshow("Results", im)


def main():
    if len(sys.argv) > 1:
        input_image = cv2.imread(sys.argv[1])
    else:
        input_image = cv2.imread("images/convoluted-webpage.png")

    qr_decoder = cv2.QRCodeDetector()
    t = time.time()
    data, bbox, rectifiedImage = qr_decoder.detectAndDecode(input_image)
    print("Time taken for Detect and Decode : {:.3f} seconds".format(
        time.time() - t))

    if len(data) > 0:
        print("Decoded Data: {}".format(data))
        display(input_image, bbox)
        rectifiedImage = np.uint8(rectifiedImage)
        cv2.imshow("Rectified QR Code", rectifiedImage)
    else:
        print("QR Code not detected")
        cv2.imshow("Results", input_image)

    cv2.waitKey(0)
    cv2.destroyAllWindows()


if __name__ == "__main__":
    main()
